File: glut_opengl_estudos/02_primitivas_e_sistema_de_coordenadas.cpp
/*
Aula 02 — Primitivas e sistema de coordenadas NDC (C++)

Objetivo pedagógico:
- Visualizar eixos em NDC (Normalized Device Coordinates: [-1, 1]).
- Desenhar pontos, linhas e triângulos no mesmo frame.
- Entender espessura de linha e tamanho de ponto.
*/

#include <GL/glut.h>
#include "cpp_glut_utils.hpp"

namespace {

void drawAxes() {
    glLineWidth(2.0f);
    glBegin(GL_LINES);
    glColor3f(0.7f, 0.7f, 0.7f);
    glVertex2f(-1.0f, 0.0f); glVertex2f(1.0f, 0.0f);
    glVertex2f(0.0f, -1.0f); glVertex2f(0.0f, 1.0f);
    glEnd();
}

void drawPrimitives() {
    glPointSize(10.0f);
    glBegin(GL_POINTS);
    glColor3f(1.0f, 0.9f, 0.2f);
    glVertex2f(-0.7f, 0.6f);
    glVertex2f(-0.5f, 0.45f);
    glVertex2f(-0.3f, 0.7f);
    glEnd();

    glLineWidth(4.0f);
    glBegin(GL_LINES);
    glColor3f(0.2f, 0.9f, 1.0f);
    glVertex2f(-0.1f, 0.6f); glVertex2f(0.4f, 0.75f);
    glVertex2f(-0.1f, 0.4f); glVertex2f(0.4f, 0.2f);
    glEnd();

    glBegin(GL_TRIANGLES);
    glColor3f(0.9f, 0.3f, 0.3f); glVertex2f(0.55f, 0.2f);
    glColor3f(0.3f, 0.9f, 0.4f); glVertex2f(0.9f, 0.2f);
    glColor3f(0.3f, 0.4f, 1.0f); glVertex2f(0.72f, 0.7f);
    glEnd();
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    drawAxes();
    drawPrimitives();
    glFlush();
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(900, 600);
    glutCreateWindow("Aula 02 - Primitivas e NDC (C++)");

    glClearColor(0.08f, 0.08f, 0.1f, 1.0f);
    glutDisplayFunc(display);
    glutMainLoop();
    return 0;
}

File: glut_opengl_estudos/02_primitivas_e_sistema_de_coordenadas.py
"""
Aula 02 — Primitivas e sistema de coordenadas NDC

Conceitos:
- NDC (Normalized Device Coordinates): eixo X e Y em [-1, 1].
- Primitivas básicas: pontos, linhas e triângulos.
- Espessura de linha e tamanho de ponto.

Como executar:
    python3 02_primitivas_e_sistema_de_coordenadas.py
"""

from OpenGL.GL import (
    GL_COLOR_BUFFER_BIT,
    GL_LINES,
    GL_POINTS,
    GL_TRIANGLES,
    glBegin,
    glClear,
    glClearColor,
    glColor3f,
    glEnd,
    glFlush,
    glLineWidth,
    glPointSize,
    glVertex2f,
)
from OpenGL.GLUT import GLUT_RGB, GLUT_SINGLE, glutCreateWindow, glutDisplayFunc, glutInit, glutInitDisplayMode, glutInitWindowSize, glutMainLoop


def draw_axes() -> None:
    """Desenha eixos X e Y para referência visual."""
    glLineWidth(2.0)
    glBegin(GL_LINES)
    glColor3f(0.7, 0.7, 0.7)
    glVertex2f(-1.0, 0.0)
    glVertex2f(1.0, 0.0)
    glVertex2f(0.0, -1.0)
    glVertex2f(0.0, 1.0)
    glEnd()


def draw_primitives() -> None:
    """Mostra três tipos de primitivas no mesmo frame."""
    glPointSize(10.0)
    glBegin(GL_POINTS)
    glColor3f(1.0, 0.9, 0.2)
    glVertex2f(-0.7, 0.6)
    glVertex2f(-0.5, 0.45)
    glVertex2f(-0.3, 0.7)
    glEnd()

    glLineWidth(4.0)
    glBegin(GL_LINES)
    glColor3f(0.2, 0.9, 1.0)
    glVertex2f(-0.1, 0.6)
    glVertex2f(0.4, 0.75)
    glVertex2f(-0.1, 0.4)
    glVertex2f(0.4, 0.2)
    glEnd()

    glBegin(GL_TRIANGLES)
    glColor3f(0.9, 0.3, 0.3)
    glVertex2f(0.55, 0.2)
    glColor3f(0.3, 0.9, 0.4)
    glVertex2f(0.9, 0.2)
    glColor3f(0.3, 0.4, 1.0)
    glVertex2f(0.72, 0.7)
    glEnd()


def display() -> None:
    """Desenha eixos + primitivas."""
    glClear(GL_COLOR_BUFFER_BIT)
    draw_axes()
    draw_primitives()
    glFlush()


def main() -> None:
    """Entry point da aula 02."""
    glutInit()
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
    glutInitWindowSize(900, 600)
    glutCreateWindow(b"Aula 02 - Primitivas e NDC")
    glClearColor(0.08, 0.08, 0.1, 1.0)
    glutDisplayFunc(display)
    glutMainLoop()


if __name__ == "__main__":
    main()
